GH-1: Initial implementation (resolves #1).

duhrer opened the pull request:
See #1 for details.

duhrer commented:
This is not ready for review yet, just making it visible to give interested parties a chance to look at it.

duhrer commented:
I am still working on:

1. Tests for the port connector(s).
2. At least limited tests for the UI components.
3. More demos.
4. Markdown documentation for all core grades.

amb26 commented on demos/js/passthrough.js:
```diff
@@ -0,0 +1,48 @@
+ *
+ *  Licensed under the MIT license, see LICENSE for details.
+ */
+(function (fluid) {
```
These days I am tending to do without the file-scope closure, but you may still like it - we got rid of all the ones in Infusion, primarily since they used to close over our versioned globals which then made a lot of diff noise on every update.

duhrer replied:
I don't have versioned globals and I prefer the style as long as it's not causing problems.

amb26 commented on demos/message-monitor-view.js:
```diff
@@ -0,0 +1,100 @@
+
+        distributeOptions: {
+            // TODO: This is probably, umm, a bit heavy-handed.
+            target: "{/ youme.connection}.options",
```
It's not bad, but I'd put a namespace in here so that these definitions can be located/deduplicated - also you may wish to parcel the listener into a grade rather than writing it out longhand which may create subtle expansion issues longterm

duhrer replied:
This grade hasn't been rewritten yet, I just straight up copied it over.  But I will keep it in mind when I get there.

duhrer replied:
I wrote a replacement demo and removed all the (now) unused functions.

amb26 commented on docs/webmidi-mock.md:
```diff
@@ -0,0 +1,177 @@
+| Name               | Type          | Description                                                                              |
+|--------------------|---------------|------------------------------------------------------------------------------------------|
+| `sysexEnabled`     | `Boolean`     | Whether we can send/receive sysex messages using this access object.                     |
+| `inputs`           | `Map`         | A map of input mocks (see below).                                                        |
```
Perhaps some more specific type here, a la @typedef that we sometimes get in JSDocs? E.g. Map<String, Something> ?

duhrer replied:
Fixed.

amb26 commented on docs/webmidi-mock.md:
```diff
@@ -0,0 +1,177 @@
+
+| Name    | Type       | Description                                                                             |
+|---------|------------|-----------------------------------------------------------------------------------------|
+| `send`  | `Function` | A function that handles requests to send messages to this port.                         |
```
Function something -> something?

duhrer replied:
I refactored the method docs to look more like the invoker docs.

amb26 commented on src/core.js:
```diff
@@ -0,0 +1,661 @@
+     *
+     */
+    youme.findPorts = function (ports, portSpec) {
+        var portFinder = youme.findPorts.portFinder(portSpec);
```
This is more a "note to self" than anything else - note that what we would really like to achieve is that portFinders are some kind of "memoised" function of the portSpec. This is the kind of thing that React's programming model makes quite easy. As it is, we have to recreate this garbage on every invocation of this function whether portSpec has changed or not.

amb26 commented on src/multiPortConnector.js:
```diff
@@ -0,0 +1,241 @@
+        gradeNames: ["youme.multiPortConnector.connections"],
+
+        dynamicComponents: {
+            connection : {
```
Odd stray space here, funny the linter doesn't catch it

duhrer replied:
There's a rule, but it's not enabled by default:

https://eslint.org/docs/rules/key-spacing

I've enabled it in my `.eslintrc.json` for now.  We can make a ticket for eslint-config-fluid if this is something we do consistently in a range of projects.

duhrer replied:
The defaults also enforce only a single space after the colon, which is not always my preferred style, but if I can give up ellipses in all my test messages, I can also try that.

amb26 commented on src/multiPortConnector.js:
```diff
@@ -0,0 +1,241 @@
+        model: {
+            portSpecs: [],
+            ports: "{youme.system}.model.ports"
+        },
```
This does indeed look like a classic case where lensed components would save a lot of risk and heartache. What I suggest you do is recast the "attemptConnection" function as a model relay into a fresh area of the model named "connectionSources" and then lens the "connection" dynamicComponents from "{that}.model.connectionSources". Should then let you axe the component visitor, and the extra level of components, and the "createOnEvent" annotations.

amb26 replied:
You might want to try simply destroying the enclosing component and rely on the natural destroy semantics to destroy and hence close all the children automatically.

duhrer replied:
I believe this has been addressed by the lensing stuff and modelised open/close stuff.

amb26 commented on src/ui/selectBox.js:
```diff
@@ -0,0 +1,110 @@
+        var optionsContent = "";
+
+        if (that.options.includeNone) {
+            optionsContent += fluid.stringTemplate(itemTemplate, { id: "none", name: that.options.noneLabel });
```
Again another note to self - given I find myself writing a lot of this kind of stuff these days too, it's seeming more and more plausible that Infusion 6 will render based on some embedded copy of preact.
Note that preact also now bundles a more lightweight "HTM" format for templates that we could support directly in place of the component's template option: https://preactjs.com/guide/v10/getting-started/#alternatives-to-jsx

amb26 replied:
If you like it, should be easy enough to create an override of renderContainer https://github.com/fluid-project/infusion/blob/main/src/framework/core/js/NewViewSupport.js#L152 which uses HTM bound to an h( .... ) function which creates real DOM nodes rather than fake ones - https://github.com/developit/htm#advanced-usage

duhrer replied:
I have been thinking a lot about this, but would like to do this as another small package to avoid having my non-music stuff (gamepad navigator, etc.) have to depend on this package.

amb26 commented on src/ui/selectBox.js:
```diff
@@ -0,0 +1,110 @@
+    youme.selectBox.handleSelectChange = function (that) {
+        var selectElement = that.locate("selectBox");
+        var selected = selectElement.val();
+        that.applier.change("selectedItemId", selected);
```
Should be possible to do this using a new-style relay:
````
    modelRelay: {
        bindValue: {
            source: "dom.selectBox.value",
            target: "selectedItemId"
        }
    }
````

duhrer replied:
Thanks, that's definitely a better pattern.

duhrer commented on src/connection.js:
```diff
@@ -0,0 +1,272 @@
+        },
+
+        listeners: {
+            "onCreate.startListening": {
```
I think I moved this to `onCreate` in a very sloppy fashion, it should always have been bound to `onPortOpen` instead, so that `{that}.open` does indeed offer another chance to bind the listener.

I think I messed this up when I was writing the tests, which failed if I was auto opening, because the port open event triggered during the auto open phase meant that the listener was bound too late for the associated check.  I tidied up the grade and the test.

I'll think about the other pattern, I like this because it matches the previous behaviour, letting authors decide whether to open on create or not.

duhrer replied:
I believe this is now handled cleanly, especially with the modelised open/close stuff.

amb26 commented on src/core.js:
```diff
@@ -0,0 +1,661 @@
+        youme.requestAccess(false, software, wrappedSuccess, error);
+    };
+
+    youme.createPortViews = function (portsArray) {
```
In the new world, perhaps "portsToViews" or some similar kind of name?

duhrer replied:
I used your suggested name.  I think this is only used in the port logging stuff, but still.

amb26 commented on src/core.js:
```diff
@@ -0,0 +1,661 @@
+
+    youme.createPortViews = function (portsArray) {
+        return fluid.transform(portsArray, function (port) {
+            return {
```
More directly could be fluid.filterKeys(port, ["id", "name", "manufacturer", "state", "connection"])

duhrer replied:
Done.

amb26 commented on src/core.js:
```diff
@@ -0,0 +1,661 @@
+     *
+     */
+    youme.collectPorts = function (type, access, ports) {
+        var portsForType = ports[type] = ports[type] || [],
```
Somewhat worrying that each call to this function may end up accumulating further ports in the supplied map - and also that despite being listed as read-only, the ports map is modified. As far as I could see, this effect could be had by simply shallow-cloning "access" into "ports", e.g. 

ports[type] = [...access[type].values()] 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#syntax

duhrer replied:
Done!  On a side note, it's so nice to be able to use ellipses again.

duhrer replied:
Note to self, need to review all map to array usages to confirm whether we would benefit from the spread operator in other places.

amb26 commented on src/core.js:
```diff
@@ -0,0 +1,661 @@
+
+    // Unsupported, non-API function.
+    youme.read.sysex = function (data) {
+        var begin = data[0] === 0xF0 ? 1 : 0,
```
Explanation of the purpose of the algorithm would be helpful here, perhaps with ref to MIDI specs

duhrer replied:
I added a typedef for the MIDI message data that included a link to the specification.  I didn't add it to every non-API function, but made sure the type was referenced in `youme.read` and `youme.write`.

amb26 commented on src/multiPortConnector.js:
```diff
@@ -0,0 +1,265 @@
+        },
+
+        invokers: {
+            open: {
```
You may want to replace all of these invokers and the calling machinery with a simple boolean flag in the model 
{
    open: false
}
which is relayed to all of the children, and then rely on local listeners to react to that state.

duhrer replied:
The more I thought about this on my morning ride, the more I liked the idea, as it both improves stability, and cleanly restores control over whether the connection is started, including on startup.  I made a go at this and made the required changes to the port connector grades as well.

amb26 commented on README.md:
```diff
@@ -1,3 +1,60 @@
-# youme
+# fluid-midi
```
Should be youme after all?

amb26 commented on src/multiPortConnector.js:
```diff
@@ -0,0 +1,254 @@
+                type: "youme.connection.output",
+                options: {
+                    // If use this, we only ever hear from one input, where the relay approach allows it.
+                    // listeners: {
```
This works for inputs, so presumably for outputs too?

duhrer replied:
We should pair on it, if I use the sensible commented-out relay strategy, all of the outputs only receive messages from one input.

duhrer replied:
I added tests and confirmed that it works in isolation (only inputs, only outputs) as well as in combination (inputs connected to outputs).  You'll see in the tests that I had to resort to `setTimeout` for one of the tests, because I can't wait for all dynamic children to fire a given event.  This would also have been a problem in the IoC world as far as I know.

duhrer commented on src/connection.js:
```diff
@@ -0,0 +1,272 @@
+        },
+
+        events: {
+            onPortOpen: null,
```
The port may be repeatedly opened and closed, but I think it might be useful for the `onReady` event in `youme.system`.  Are there examples of this type of usage I could refer to?

duhrer replied:
I tried this for a while and couldn't get the resource loading to block `onCreate`.  [Here's a pen](https://codepen.io/duhrer/pen/BaryBQv) with an example, in all cases `onCreate` fires before the promise is resolved.

duhrer replied:
I updated the example, it seems like onCreate does wait for the promise to resolve, but always beats anything else "thenned" to the same promise.

duhrer replied:
Ah, I see, it's down to the order in which they were registered.  I will give that some thought and work on the timing tomorrow.

duhrer commented:
This is coming along, but as the conversation is getting longer, a quick summary.  In addition to whatever else Antranig and Colin would like to see addressed, I still need to complete:

1. The demos (I want to add a "MIDI pipes" demo).
2. The docs (I need to document the UI grades).

I also want to cut a dev release and try using it in another project.

amb26 commented on demos/js/messageMonitor.js:
```diff
@@ -0,0 +1,161 @@
+                                        options: {
+                                            // Listen at this level so that we can determine which port/device sent the message.
+                                            listeners: {
+                                                "onMessage.log": {
```
Looks like a great candidate for being parcelled up in a grade which is then sent with an options distribution

duhrer replied:
Done, much cleaner!

amb26 commented on demos/js/messageMonitor.js:
```diff
@@ -0,0 +1,161 @@
+        messageMonitor.applier.change(["loggedMessages", index], toLog);
+    };
+
+    youme.demos.messageMonitor.resetLog = function (messageMonitor) {
```
There is now a standard utility
````
fluid.replaceModelValue(messageMonitor.applier, [], [])
````
which will do this in one step

duhrer replied:
Cheers, I went through the code and replaced the pattern with the new function throughout.  I left the transactions (in the tests) which only delete the value alone.

amb26 commented on src/core.js:
```diff
@@ -0,0 +1,655 @@
+     *
+     * @param {"inputs"|"outputs"} type - The type of inputs, only "inputs" and "outputs" are supported.
+     * @param {MIDIAccess} access - The `MIDIAccess` object provided by the WebMIDI API.
+     * @param {Map<MIDIPort>} ports - A read-only map-like structure containing the available ports for `type`.
```
Nicely cleaned up, however, remember to remind the user that this "read-only" structure is actually modified!
Since the only caller of collectPorts is getPorts, and this function is now a one-liner, why not abolish it and just write the spreads directly in getPorts?

duhrer replied:
Done.

amb26 commented on src/ui/multiSelectBox.js:
```diff
@@ -0,0 +1,102 @@
+    youme.multiSelectBox.renderOptions = function (that, itemTemplate) {
+        var optionsContent = "";
+        fluid.each(that.model.optionItems, function (optionItem) {
+            var selected = that.model.selectedItemIds.indexOf(optionItem.id) !== -1 ? "selected" : "";
```
Modern is https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

duhrer replied:
Done, and searched for other uses throughout.

amb26 commented on src/ui/portSelectorView.js:
```diff
@@ -0,0 +1,143 @@
+        }
+    });
+
+    youme.portSelectorView.updatePortSpec = function (that) {
```
replaceModelValue again here

amb26 commented on tests/js/multiPortConnector-tests.js:
```diff
@@ -0,0 +1,398 @@
+            midiSystem: {
+                options: {
+                    listeners: {
+                        "onReady.runTests": {
```
Another benefit of the "creation is readiness" pattern is that you could use the onCreate promisified event as the return value from promisified test fixtures

duhrer commented on tests/js/portConnector-tests.js:
```diff
@@ -0,0 +1,257 @@
+ */
+/*global jqUnit */
+// TODO: These have to be run last or they will cause problems with subsequent tests.
+// I suspect additional listener calls or faulty start/stop logic are to blame.
```
If we end up having to have many of these things, I'll keep this note in mind.  For now, I'm OK with just running it last and leaving the TODO.

duhrer commented:
Here's the error, which occurs while the portConnector's connection is being constructed.

```
TypeError: Cannot read properties of undefined (reading 'initModelTransactionId')
    at fluid.enlistModelComponent (http://localhost:7357/node_modules/infusion/dist/infusion-all.js:148:122208)
    at fluid.establishModelRelay (http://localhost:7357/node_modules/infusion/dist/infusion-all.js:148:146922)
    at fluid.expander.invokeFunc [as typeFunc] (http://localhost:7357/node_modules/infusion/dist/infusion-all.js:148:116302)
    at fluid.expandImmediateImpl (http://localhost:7357/node_modules/infusion/dist/infusion-all.js:148:113546)
    at fluid.expandImmediate (http://localhost:7357/node_modules/infusion/dist/infusion-all.js:148:113301)
    at fluid.memberFromRecord (http://localhost:7357/node_modules/infusion/dist/infusion-all.js:148:26913)
    at Array.strategy (http://localhost:7357/node_modules/infusion/dist/infusion-all.js:148:28140)
    at fluid.model.traverseWithStrategy (http://localhost:7357/node_modules/infusion/dist/infusion-all.js:148:117256)
    at fluid.model.accessImpl (http://localhost:7357/node_modules/infusion/dist/infusion-all.js:139:15457)
    at fluid.model.accessWithStrategy (http://localhost:7357/node_modules/infusion/dist/infusion-all.js:148:118497)
```

amb26 commented:
Thanks, this is a framework bug which I've written up as https://issues.fluidproject.org/browse/FLUID-6741

duhrer commented:
@amb26, The tests are working again with an attempt at moving to 'create is ready' for key components.  This is ready for further review.

amb26 commented on demos/js/messageMonitor.js:
```diff
@@ -0,0 +1,146 @@
+
+        invokers: {
+            resetLog: {
+                funcName: "fluid.replaceModelValue",
```
Fits pretty neatly now!

amb26 commented on docs/ui-components.md:
```diff
@@ -0,0 +1,168 @@
+    },
+    modelListeners: {
+        label: {
+            // Exclude "init" so that we don't interfere with the initial render.
```
Reminder to self to make an "html" materialiser here. Is there any harm to missing off excludeSource: init? It seems that if the label isn't in the document yet the jQuery action will just be a noop?

duhrer replied:
I think I was worried about duplicate render cycles, but everything works without it, so I got rid of them in the select box, multi select box, and these docs.

amb26 commented on src/js/multiPortConnector.js:
```diff
@@ -0,0 +1,139 @@
+        },
+
+        modelListeners: {
+            "ports": {
```
These might be better written using the "listen on multiple paths" syntax - https://docs.fluidproject.org/infusion/development/changeapplierapi#the-path-entry-in-a-model-listener-declaration
i.e. with a member path: ["ports", "portSpecs"] and then you can use the key for a namespace, e.g. findPorts

amb26 replied:
Or alternatively simply as a modelRelay specification which will take care of invalidation for you,e.g.
````
modelRelay: {
    findPorts: {
        target: "connectionPorts",
        func: "youme.multiPortConnector.findPorts",
        args: ["{that}", "{that}.options.direction", "{that}.model.ports", "{that}.model.portSpecs"]
    }
}
````

duhrer replied:
If I use the suggested code (which sounds great, BTW), I get circular reference errors, as in:

```
Error: Assertion failure - check console for more details: Error in component configuration - a circular reference was found during evaluation of path segment "model of path ["model"]": for more details, see the activity records following this message in the console, or issue fluid.setLogging(fluid.logLevel.TRACE) when running your application
```

duhrer replied:
Sadly, the multi-path relay doesn't seem to work either, which is not too surprising, as it seems to longer be tested.  Here's [a codepen demonstrating the lack of relay](https://codepen.io/duhrer/pen/abYmzKy).

amb26 replied:
> Sadly, the multi-path relay doesn't seem to work either, which is not too surprising, as it seems to longer be tested. Here's [a codepen demonstrating the lack of relay](https://codepen.io/duhrer/pen/abYmzKy).

It is tested here - https://github.com/fluid-project/infusion/blob/main/tests/framework-tests/core/js/DataBindingTests.js#L1923-L1927 the correct syntax is to use "path" which works in the codepen

amb26 replied:
> If I use the suggested code (which sounds great, BTW), I get circular reference errors, as in:

Could you check in a branch for me to investigate, cheers

duhrer replied:
That's also why I thought there were no tests, as I was searching for `paths`.

duhrer replied:
Or maybe not, I don't see any usages of array `path` values either in the tests, but it does at least work.

amb26 replied:
>> If I use the suggested code (which sounds great, BTW), I get circular reference errors, as in:
> Could you check in a branch for me to investigate, cheers

I found this works fine if the relay definition is written to be a proper side-effect-free function as follows:

````
    youme.multiPortConnector.findPorts = function (that, direction, ports, portSpecs) {
        // Create it as a map indexed by ID so that we can avoid duplicates.
        var connectionPortsById = {};
        fluid.each(portSpecs, function (portSpec) {
            var portsToSearch = fluid.makeArray(ports && ports[direction]);
            if (portsToSearch.length > 0) {
                var foundPorts = youme.findPorts(portsToSearch, portSpec);
                fluid.each(foundPorts, function (port) {
                    connectionPortsById[port.id] = port;
                });
            }
        });
        var connectionPorts = Object.values(connectionPortsById);
        return connectionPorts;
    };
````

Note that you can't issue explicit reads to the model from inside a relay definition, these all need to be hoisted into arguments so that invalidation can properly be checked.

duhrer replied:
Thanks, that works.  I'll give the port connector another look shortly.

amb26 commented on src/js/core.js:
```diff
@@ -0,0 +1,616 @@
+     *
+     * @typedef MIDIMessageData
+     * @type {Uint8Array}
+     * @return {Promise} - A promise that will resolve when access is granted or reject otherwise.
```
Should explain what it will resolve to - looks like it is a native MIDIAccess object?

duhrer replied:
Yes, I added a comment to that effect.

amb26 commented on src/js/system.js:
```diff
@@ -0,0 +1,122 @@
+        },
+
+        model: {
+            access: "{that}.resources.access.parsed",
```
Should explain what the resulting type of this is, is it the native MIDIAccess object?

duhrer replied:
I added a comment.

duhrer commented:
@amb26, I believe I've addressed all the comments in the latest review.  I left the non-working model relay in the multi port connector commented out for your review, other than that this is good to go as far as I can tell.

amb26 commented on tests/html/portConnector-tests.html:
```diff
@@ -0,0 +1,44 @@
+        <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
+        <title>Port Connector Tests</title>
+
+        <link rel="stylesheet" media="screen" href="../../../node_modules/infusion/tests/lib/qunit/css/qunit.css" />
```
Wrong number of ../ here

amb26 replied:
And in all other test drivers

duhrer replied:
Fixed, throughout.

amb26 commented on demos/js/console.js:
```diff
@@ -0,0 +1,258 @@
+                    args: ["{that}.model.errorString"]
+                },
+                {
+                    this: "{that}.dom.error",
```
This could be a new-style materialised relay:
modelRelay: {
    errorStyle: {
        target: "dom.error.class.console-visible",
        source: "{that}.model.errorString"
    }
}

amb26 commented on demos/js/console.js:
```diff
@@ -0,0 +1,258 @@
+        modelListeners: {
+            errorString: [
+                {
+                    excludeSource: "init",
```
This could be a new-style materialised relay:
````
modelRelay: {
    errorText: {
        target: "dom.error.text",
        source: "{that}.model.errorString"
    }
}
````

duhrer replied:
Done.

amb26 commented on src/js/portConnector.js:
```diff
@@ -0,0 +1,172 @@
+        },
+        modelListeners: {
+            "ports": {
+                // excludeSource: "init",
```
One of these is commented out and the other not, presumably both can go. I think this could be replaced by a relay just as we can in multiPortConnector

duhrer replied:
I knew both of them would be set on startup and wanted to avoid a double bump, but either a multi-path listener or relay sound like better options.

duhrer replied:
I realised before posting the question about multi-path model relays that model relays run whenever any of their args are updated, so I was able to make this work.

amb26 commented on src/js/connection.js:
```diff
@@ -0,0 +1,256 @@
+        gradeNames: ["fluid.modelComponent"],
+
+        model: {
+            port: false,
```
Some type instability here in that it seems that when open this holds a MIDIPort - probably null is a better initial value, and also document this type in the model

duhrer replied:
I added a note here and updated the initial value.

amb26 commented on src/js/connection.js:
```diff
@@ -0,0 +1,256 @@
+
+    youme.connection.handlePortChange = function (oldPort, newPort, isOpen, onPortOpen, onPortClose, onError) {
+        if (oldPort) {
+            oldPort.close();
```
Looks like there is a risk of a race here since the port closing operation is asynchronous and the promise return value is lost here. It seems best to turn handlePortChange into a transforming promise chain which closes the old port in one listener and opens the new one in another. Failing that, this can become a regular async function.

duhrer replied:
I ended up having to use some of the old conventions to ensure that rejection would fire `onError`, but did at least line these up as links in a chain and ensure that they return promises.

amb26 commented on src/js/connection.js:
```diff
@@ -0,0 +1,256 @@
+    youme.connection.updateConnectionState = function (port, isOpen, onPortOpen, onPortClose, onError) {
+        if (port) {
+            if (isOpen) {
+                youme.connection.open(port, onPortOpen, onError);
```
Should be async function or transforming chain

duhrer commented on demos/js/console.js:
```diff
@@ -0,0 +1,258 @@
+                }
+            ],
+            editorString: {
+                excludeSource: "init",
```
Also, this is the only one that didn't do what I expected when converting to a `modelRelay`.  The initial value is empty.

duhrer replied:
I pushed up what I have, you can see that the component (renamed to `midiConsole` to avoid the previous goof of clobbering the global `console`) has the editor strings in its model, and that the individual components each have their editor string, but nothing is displayed.

amb26 replied:
Just checking on the status of this - https://issues.fluidproject.org/browse/FLUID-6746 got accidentally closed due to a typo in the commits for https://github.com/fluid-project/infusion-docs/commit/b37297e84c8964e81929c7f2317ed8f49b8a398c to infusion-docs which swapped FLUID-6476 for FLUID-6746 - FLUID-6746 should be listed as still outstanding as highlighted by this relay problem

duhrer commented on src/js/connection.js:
```diff
@@ -0,0 +1,256 @@
+    };
+
+    youme.connection.open = function (port, onPortOpen, onError) {
+        var portOpenPromise = port.open();
```
Done.

duhrer commented:
In addition to whatever concerns you have, we need to figure out why the modelRelay for `editorString` doesn't work intiially.

amb26 commented on src/js/ui/multiPortSelectorView.js:
```diff
@@ -0,0 +1,168 @@
+        }
+    });
+
+    youme.multiPortSelectorView.findDesiredPorts = function (that) {
```
This one can also be converted from the effectful modelListener to modelRelay style

duhrer replied:
Done.

amb26 commented:
Cheers @duhrer this looks good to go! :1st_place_medal:

colinbdclark commented on docs/core-components.md:
```diff
@@ -0,0 +1,201 @@
+| `onSysex`       | An event that is fired when a "sysex" message is received.          |
+| `onTuneRequest` | An event that is fired when a "tune request" message is received.   |
+
+See [the WebMIDI API documentation](https://www.midi.org/specifications-old/item/table-1-summary-of-midi-message) for
```
This looks like it links to the MIDI specification, rather than the WebMIDI API documentation as the link text suggests.

duhrer replied:
Yes, the WebMIDI API links to that rather than spell it out on its own, so I changed it to "the MIDI specificiation".

colinbdclark reviewed (commented):
Sorry for the long delay in submitting these very minor comments. Everything looks great in this PR, and my comments are suggestions only. Thanks so much for your hard work on this, @duhrer.

duhrer commented on docs/core-components.md:
```diff
@@ -0,0 +1,201 @@
+
+### Port Specifications
+
+A port specification is an object that describes the port to which we wish to connect. Consists of or more of the
```
Cheers, fixed.

duhrer commented on docs/index.md:
```diff
@@ -0,0 +1,17 @@
+
+## Core Components
+
+The core components in YouMe can be used to request access via the WebMIDI API, to interact with ports, and to
```
Good suggestion, thanks.

duhrer commented on docs/core-components.md:
```diff
@@ -0,0 +1,201 @@
+
+In addition to the events for the base `youme.connection` grade, this grade has the following events.
+
+| Event             | Description                                                |
```
That seems reasonable, I'll give some thought to where to position the overview and refactor that shortly.

duhrer replied:
I made an attempt at this, please review and let me know if you have any input.

duhrer commented:
In testing this work downstream, I discovered a timing issue with sending messages before a component had fully initialised.  It's a problem that only shows up when running Infusion 4.5.0 in that package.  I updated this package to use 4.5.0, and the tests still pass.  My guess is it was a timing issue with one of my "lazy that-ist" functions that just passes `that` and then drills down into `options` from there.

I updated the relevant invoker to pass the required information using fully specified IoC references.  It works under Infusion 4.0.0 as well as 4.5.0, but I'd rather leave the required Infusion version for youme set to 4.5.0, as I want to guard against similar errors.

colinbdclark commented:
I'll defer to @amb26 about the fix for the timing issues. From my perspective, everything is looking super good on this now, and I'm excited to start using it.

duhrer commented:
I looked through the rest of the code base, there are a few lingering "lazy that-ist" patterns, all in the tests and all are working without issues.  There are none left in the core code.

colinbdclark commented:
LGTM. I think we can merge this and if there are any remaining design issues related to the that-ist pattern, we can address those under a separate ticket.
